dspace-api/src/main/java/org/dspace/app/orcid/service/impl/OrcidSynchronizationServiceImpl.java

```diff
@@ -0,0 +1,258 @@
+/**
+ * The contents of this file are subject to the license and copyright
+ * detailed in the LICENSE and NOTICE files at the root of the source
+ * tree and available online at
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.orcid.service.impl;
+
+import static java.time.LocalDateTime.now;
+import static java.time.format.DateTimeFormatter.ISO_DATE_TIME;
+import static java.util.List.of;
+import static org.apache.commons.lang3.EnumUtils.isValidEnum;
+import static org.apache.commons.lang3.StringUtils.isBlank;
+import static org.dspace.content.Item.ANY;
+
+import java.sql.SQLException;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.apache.commons.codec.binary.StringUtils;
+import org.dspace.app.orcid.model.OrcidEntityType;
+import org.dspace.app.orcid.model.OrcidTokenResponseDTO;
+import org.dspace.app.orcid.service.OrcidSynchronizationService;
+import org.dspace.app.profile.OrcidEntitySyncPreference;
+import org.dspace.app.profile.OrcidProfileDisconnectionMode;
+import org.dspace.app.profile.OrcidProfileSyncPreference;
+import org.dspace.app.profile.OrcidSynchronizationMode;
+import org.dspace.authorize.AuthorizeException;
+import org.dspace.content.Item;
+import org.dspace.content.MetadataValue;
+import org.dspace.content.service.ItemService;
+import org.dspace.core.Context;
+import org.dspace.eperson.EPerson;
+import org.dspace.eperson.service.EPersonService;
+import org.dspace.services.ConfigurationService;
+import org.springframework.beans.factory.annotation.Autowired;
+
+/**
+ * Implementation of {@link OrcidSynchronizationService}.
+ *
+ * @author Luca Giamminonni (luca.giamminonni at 4science.it)
+ *
+ */
+public class OrcidSynchronizationServiceImpl implements OrcidSynchronizationService {
+
+    @Autowired
+    private ItemService itemService;
+
+    @Autowired
+    private ConfigurationService configurationService;
+
+    @Autowired
+    private EPersonService ePersonService;
+
+    @Override
+    public void linkProfile(Context context, Item profile, OrcidTokenResponseDTO token) throws SQLException {
+
+        String orcid = token.getOrcid();
+        String accessToken = token.getAccessToken();
+        String refreshToken = token.getRefreshToken();
+        String[] scopes = token.getScopeAsArray();
+
+        itemService.setMetadataSingleValue(context, profile, "person", "identifier", "orcid", null, orcid);
+        itemService.setMetadataSingleValue(context, profile, "dspace", "orcid", "access-token", null, accessToken);
```

Comment:
@LucaGiamminonni  & @benbosman :  Talked briefly with @abollini about this point today & we came to a compromise solution which I hope could work here.  

Together we propose that these new ORCID metadata fields be **hardcoded** into the Java code for `metadata.hide.*`.  So, they will still be stored in metadata & use `metadata.hide` approach, but is should NOT be possible to "un-hide" them & they won't appear in `dspace.cfg`.  Additionally, we should review the ITs for `metadata.hide` to ensure they are sufficiently testing security & ensuring those metadata fields don't ever get indexed into Solr (for discovery or OAI or anything else).

As we know, this is the first instance of storing highly secure information in metadata fields.  I see the reason why this may be ideal, as it allows this information to "live" with the object, without constantly requiring new database tables/columns -- and there's plenty of future ways this could become useful.

For 7.4, I'd also recommend a more thorough review of this practice & ensure `metadata.hide` is as secure as it can possibly be (I think it likely is, but as I said...this is a slight change in what we store in metadata, so it's worth the additional deep check.)

Reply:
I very strongly disagree with this. We should at the very least encrypt this data. The chances for data leaks, the full access for admins, … is too risky to store tokens
Encrypting this data should be a small fix to provide some security

Reply:
@abollini or @LucaGiamminonni : See @benbosman 's note above. Is there any way to do some sort of encryption of these metadata field values for better security? 

I realize I overlooked something that @benbosman has noticed...even with `metadata.hide` settings in place, any DSpace Administrator can still read all these metadata fields.  This seems like it'd be a concern, as connecting your ORCID to DSpace shouldn't give DSpace Site Administrators full access to make any modifications to your external ORCID Profile....you are only agreeing to allow the _DSpace system_ to sync data to/from your ORCID profile.

Perhaps we could find a simple way to encrypt this data using the same algorithm/salt we already store in the `EPerson` for encrypting passwords?  Essentially, maybe there's a way we could just use the existing `org.dspace.eperson.PasswordHash` to encrypt these field values? 

That'd hopefully keep the code changes minimal, and ensure only the EPerson can decrypt this metadata field's value (as it'd be hashed in the same way as an EPerson's password).

Reply:
I also am wary of storing cleartext tokens.  But where are we to get the key material with which to decrypt them?  (I need to study the flow of this code.  If the user is present when the tokens are used then we should have access to his secrets.)

The password digestion process is intentionally one-way.  To recover a usable token, we must use a reversible encryption such as AES, not a hash.  Password validation is done by comparing hashes; the original password is only recoverable by an infeasible amount of work.  Hashes would only serve if ORCiD possesses the identical hashes, which I think could only be accomplished securely by _e.g._ a DH key agreement dialog.  Do they offer that?

I worry about using "hidden" field names.  Too many opportunities for unintended disclosure.  Notice that anything stored will be present in database dumps, and therefore your backup tapes.

Reply:
@mwoodiupui : Oh, good point. I obviously wasn't thinking through how the PasswordHash works & the fact that it's one-way.  So, it wouldn't work in this scenario.

In the scenario, the ORCID tokens (which allow full access to read/write to one's ORCID acct) are stored in DSpace because the user is not present all the times they are needed.  Specifically, there's a "sync" option (in the follow PR #8266 ) which uses these ORCID tokens to sync your DSpace data (in your Researcher Profile, see #8232) to/from your ORCID acct on a regular basis.  So, when the user logs into DSpace via ORCID, they are essentially connecting their ORCID acct to their DSpace EPerson to allow the two to share/sync information.

So, you are correct here we'd need to use a reversible encryption, as it'd need to be something DSpace can do without the user being actively logged into the system.

Reply:
@LucaGiamminonni and @abollini :  Please see the discussion above.  I can see both @benbosman 's and @mwoodiupui 's points here.  So, that means we have 3 Committers (self included) who all have concerns with the current approach. We need to find an alternative way forward... a few immediate options (there may be others)
* Keeping this in "hidden" metadata fields & encrypting this data (so that even though this data is accessible to Admins in the UI, they cannot use/decrypt it), **or** 
* Moving it out of metadata fields into a separate database column (making it inaccessible to _anyone_ from the UI / REST API). Arguable though, we still may want to encrypt it, so that it's not plaintext in database dumps/backups.

If anyone sees other options, please do speak up.

Reply:
If we *did* encrypt the tokens, then we still have the question of how the decryption key is provided.  We can't use the user's password if the user is not present or an external authentication method is chosen.  If it is stored (in configuration, for example) then that is equivalent to no encryption, because anyone who can get a token can get the key.  Usually the only way out of this trap is to have an operator enter the key at startup, which is so cumbersome that most sites would probably find a way to defeat it.  Storing tokens in a table which is not remotely accessible may indeed be the best approach.

Reply:
Resolving this discussion as it's moved to a table.  The question about GDPR warnings was discussed in a meeting & we decided there's nothing DSpace needs to do as ORCID will already warn the user what information will be shared with DSpace on login.  The only thing we will need to ensure is that this is included in the (eventual) default privacy statement in #1080